source/opt/ssa_rewrite_pass.h

```diff
@@ -0,0 +1,237 @@
+  // |bb|.  Phi candidates are potential Phi instructions that will be created
+  // at the end of the SSA rewrite pass.
+  //
+  // Since the rewriting algorithm may remove Phi candidates when it finds them
+  // to be trivial, we avoid the expense of creating actual Phi instructions by
+  // keeping a pool of Phi candidates (|phi_candidates_|) during rewriting.
+  //
+  // Once the candidate Phi is created, it returns its ID.
+  PhiCandidate& CreatePhiCandidate(uint32_t var_id, ir::BasicBlock* bb);
+
+  // Adds operands to the Phi candidate |phi_cand|.  This may result in
+  // |phi_cand| becoming a trivial Phi instruction.  When that happens,
+  // |phi_cand| will get replaced by one of its operands.  This function will
+  // return either the ID for |phi_cand| or the ID of the operand that replaced
+  // it.
+  uint32_t AddPhiOperands(PhiCandidate* phi_cand);
+
+  // Attempts to remove a trivial Phi candidate |phi_cand|. Trivial Phis are
+  // those that only reference themselves and one other value |val| any number
+  // of times. This will try to remove any other Phis that become trivial after
+  // |phi_cand| is removed.
+  //
+  // If |phi_cand| is trivial, it returns the SSA ID for the value that should
+  // replace it.  Otherwise, it returns the SSA ID for |phi_cand|.
+  uint32_t TryRemoveTrivialPhi(PhiCandidate* phi_cand);
+
+  // Completes |phi_cand| by adding missing arguments or replacing unknown
+  // arguments with their correct reaching definition.
+  void CompletePhiCandidate(PhiCandidate* phi_cand);
+
+  // Finalizes processing of Phi candidates.  Once the whole function has been
+  // scanned for loads and stores, the CFG will still have some incomplete and
+  // trivial Phis.  This will add missing arguments and remove trivial Phi
+  // candidates.
+  void FinalizePhiCandidates();
+
+  // Map holding the value of every SSA-target variable at every basic block
+  // where the variable is stored. defs_at_block_[block][var_id] = val_id
+  // means that there is a store instruction for variable |var_id| at basic
+  // block |block| with value |val_id|.
+  BlockDefsMap defs_at_block_;
+
+  // Map, indexed by basic block, holding all the Phi candidates created at that
+  // basic block during SSA rewriting.  phi_candidates_[block] returns a map,
+  // indexed by variable ID, for the Phi candidate created for that variable at
+  // |block|.
+  BlockPhisMap phi_candidates_;
+
+  // SSA replacement table.  This maps variable IDs, resulting from a load
+  // operation, to the value IDs that will replace them after SSA rewriting.
+  // After all the rewriting decisions are made, a final scan through the IR is
+  // done to replace all uses of the original load ID with the value ID.
+  std::unordered_map<uint32_t, uint32_t> load_replacement_;
+
+  // Set of blocks that have been sealed already.
+  std::unordered_set<ir::BasicBlock*> sealed_blocks_;
+
+  // Memory pass requesting the SSA rewriter.
+  MemPass* pass_;
```

Comment:
A MemPass containing another MemPass?  Is this so you can get some of the unilities in MemPass?  Which ones, and do they belong there?  I would expect it to store a context, and that is it.

Reply:
No, the rewriter is a standalone class.  It can be used by any pass as a helper.  The SSARewriterPass is the pass that uses it.  I organized it this way so that I could use it from the local multi-store elimination pass.

Reply:
Don't bother to change this.  I can be done later if we want to do something about it.

Reply:
We also access other pass functions (CollectTargetVars, GetUndefVal, etc).  It can't just use the context.